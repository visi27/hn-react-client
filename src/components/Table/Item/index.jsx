import React from 'react';
import { PropTypes } from 'prop-types';
import Button from '../../Button/presenter';
import * as style from './style';

const Item = ({ item, onDismiss }) => (
  <div key={item.objectID} className="container-fluid">
    <span style={style.largeColumn}>
      <a href={item.url}>{item.title}</a>
    </span>
    <div className="list-inline">
      <span className="list-inline-item">Author: {item.author}</span>
      <span className="list-inline-item">Comments: {item.num_comments}</span>
      <span className="list-inline-item">Points: {item.points}</span>
      <span className="list-inline-item">
        <Button className="btn btn-outline-danger btn-sm" onClick={() => onDismiss(item.objectID)}>
          Dismiss
        </Button>
      </span>
    </div>
  </div>
);

Item.propTypes = {
  item: PropTypes.shape({
    created_at: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    title_highlighted: PropTypes.string,
    url: PropTypes.string,
    author: PropTypes.string.isRequired,
    points: PropTypes.number.isRequired,
    story_text: PropTypes.string,
    num_comments: PropTypes.number.isRequired,
    objectID: PropTypes.string.isRequired,
  }).isRequired,
  onDismiss: PropTypes.func,
};

Item.defaultProps = {
  onDismiss: () => {},
};

export default Item;
