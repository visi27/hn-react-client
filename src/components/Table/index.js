import { connect } from 'react-redux';
import Table from './presenter';

function mapStateToProps(state) {
  const { result } = { ...state };
  return {
    result,
  };
}

export default connect(mapStateToProps, null)(Table);
