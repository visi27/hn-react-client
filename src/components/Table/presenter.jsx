import React from 'react';
import { PropTypes } from 'prop-types';
import Item from './Item';
import conf from '../../config';
import Paginator from '../Paginator/index';

const Table = ({ result }) => (
  <div className="container-fluid">
    {result.hits.map(item => <Item key={item.objectID} item={item} />)}
    <Paginator />
  </div>
);

Table.propTypes = {
  result: PropTypes.shape({
    hits: PropTypes.array,
    page: PropTypes.number,
    nbPages: PropTypes.number,
    searchTerm: PropTypes.string,
  }),
};

Table.defaultProps = {
  result: {
    hits: [],
    page: 0,
    nbPages: 0,
    searchTerm: conf.DEFAULT_QUERY,
  },
};

export default Table;
