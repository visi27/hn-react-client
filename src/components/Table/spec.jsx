import React from 'react';
import ReactDom from 'react-dom';
import renderer from 'react-test-renderer';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Table from './presenter';
import conf from '../../config';

configure({ adapter: new Adapter() });

describe('Table', () => {
  const props = {
    result: {
      hits: [
        {
          title: '1',
          author: '1',
          num_comments: 1,
          points: 2,
          objectID: 'y',
          created_at: '2018-04-03T09:35:47.000Z',
        },
        {
          title: '2',
          author: '2',
          num_comments: 1,
          points: 2,
          objectID: 'z',
          created_at: '2018-04-03T09:35:47.000Z',
        },
      ],
      page: 0,
      nbPages: 0,
      searchTerm: conf.DEFAULT_QUERY,
    },
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDom.render(<Table {...props} />, div);
  });

  test('has valid snapshot', () => {
    const component = renderer.create(<Table {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items on the list', () => {
    const element = mount(<Table {...props} />);
    expect(element.find('.item').length).toBe(2);
  });
});
