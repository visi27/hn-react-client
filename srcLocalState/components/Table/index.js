import React, { Component } from 'react';
import Item from '../Item/index';
import Header from '../Header/index';
import PropTypes from 'prop-types';
import { sortBy } from 'lodash';

const SORTS = {
  NONE: list => list,
  TITLE: list => sortBy(list, 'title'),
  AUTHOR: list => sortBy(list, 'author'),
  COMMENTS: list => sortBy(list, 'num_comments').reverse(),
  POINTS: list => sortBy(list, 'points').reverse(),
};

class Table extends Component {
  constructor (props) {
    super(props);

    this.state = {
      sortKey: 'NONE',
      isSortReversed: false,
    };

    this.onSort = this.onSort.bind(this);
  }

  render () {
    const {list, onDismiss} = this.props;
    const {sortKey, isSortReversed} = this.state;

    const sortedList = isSortReversed && (sortKey !== 'NONE') ? SORTS[sortKey](list) : SORTS[sortKey](list).reverse();
    return (
      <div className="table">
        <Header onSort={this.onSort} activeSort={sortKey}/>
        {sortedList.map(item => (
          <Item key={item.objectID} item={item} onDismiss={onDismiss}/>
        ))}
      </div>
    );
  }

  onSort (newSortKey: string) {
    const {sortKey, isSortReversed} = this.state;

    // If new sort key is the same as old sort key treat the click as reversing sort direction.
    const shouldSortRevers = (sortKey === newSortKey) ? !isSortReversed : isSortReversed;
    this.setState({sortKey: newSortKey, isSortReversed: shouldSortRevers});
  }
}

Table.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      objectID: PropTypes.string.isRequired,
      url: PropTypes.string,
      title: PropTypes.string,
      author: PropTypes.string,
      num_comments: PropTypes.number,
      points: PropTypes.number,
    }).isRequired,
  ),
  onDismiss: PropTypes.func.isRequired,
};

export default Table;
