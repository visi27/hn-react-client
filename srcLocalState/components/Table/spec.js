import React from 'react';
import ReactDom from 'react-dom';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Table from './index';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('Table', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z' }
    ],
    onDismiss: () => {},
    onSort: () => {},
    sortKey: 'NONE',
    isSortReversedL: false
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDom.render(<Table {...props} />, div);
  });

  test('has valid snapshot', () => {
    const component = renderer.create(<Table {...props} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items on the list', () => {
    const element = mount(<Table {...props} />);
    expect(element.find('.table-row').length).toBe(2);
  });

  it('shows one header on the list', () => {
    const element = mount(<Table {...props} />);
    expect(element.find('.table-header').length).toBe(1);
  });
});
