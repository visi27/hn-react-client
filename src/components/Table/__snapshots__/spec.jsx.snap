// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table has valid snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  />
  <div
    className="container-fluid item"
  >
    <div
      className="list-inline"
    >
      <span
        className="list-inline-item"
      >
        <a
          className="h5 text-muted"
          href={undefined}
        >
          1
        </a>
      </span>
      <span
        className="list-inline-item"
      >
        <button
          className="btn btn-outline-danger btn-sm"
          onClick={[Function]}
          type="button"
        >
          Dismiss
        </button>
      </span>
    </div>
    <div
      className="list-inline"
    >
      <span
        className="list-inline-item"
      >
        Author: 
        1
      </span>
      <span
        className="list-inline-item"
      >
        Created: 
        20 days ago
      </span>
      <span
        className="list-inline-item"
      >
        Comments: 
        1
      </span>
      <span
        className="list-inline-item"
      >
        Points: 
        2
      </span>
      <span
        className="list-inline-item"
      >
        <a
          className="text-muted"
          href={undefined}
        >
          (
          )
        </a>
      </span>
    </div>
    <hr />
  </div>
  <div
    className="container-fluid item"
  >
    <div
      className="list-inline"
    >
      <span
        className="list-inline-item"
      >
        <a
          className="h5 text-muted"
          href={undefined}
        >
          2
        </a>
      </span>
      <span
        className="list-inline-item"
      >
        <button
          className="btn btn-outline-danger btn-sm"
          onClick={[Function]}
          type="button"
        >
          Dismiss
        </button>
      </span>
    </div>
    <div
      className="list-inline"
    >
      <span
        className="list-inline-item"
      >
        Author: 
        2
      </span>
      <span
        className="list-inline-item"
      >
        Created: 
        20 days ago
      </span>
      <span
        className="list-inline-item"
      >
        Comments: 
        1
      </span>
      <span
        className="list-inline-item"
      >
        Points: 
        2
      </span>
      <span
        className="list-inline-item"
      >
        <a
          className="text-muted"
          href={undefined}
        >
          (
          )
        </a>
      </span>
    </div>
    <hr />
  </div>
</div>
`;
